actionpack/lib/action_dispatch/middleware/host_authorization.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require "action_dispatch/http/request"
+
+module ActionDispatch
+  # This middleware guards from DNS rebinding attacks by white-listing the
+  # hosts a request can be sent to.
+  #
+  # When a request comes to an unauthorized host, the +response_app+
+  # application will be executed and rendered. If no +response_app+ is given, a
+  # default one will run, which responds with +403 Forbidden+.
+  class HostAuthorization
+    class SpoofProtection # :nodoc:
```

Comment:
@matthewd, I was thinking of extracting this check into `ActionDispatch::Request#remote_host{,=}` pair of methods, but decided to withhold it for now. The semantics can be similar to `ActionDispatch::Request#remote_ip{,=}`, and a spoof error can be risen if that's detected in application code. Does anyone has good examples of needing `ActionDispatch::Request#host` in straight applications?